### PR TITLE
8281822: Test failures on non-DTrace builds due to incomplete DTrace* flags handling

### PR DESCRIPTION
**Pull request by shipilev**
A test failure reproduces in GHA in x86_32 mode. Reproduces locally too:

```
$ CONF=linux-x86-server-fastdebug make run-test TEST=serviceability/7170638/SDTProbesGNULinuxTest.java

java.lang.Error: java.lang.RuntimeException: '.note.stapsd' missing from stdout

at SDTProbesGNULinuxTest.testLibJvm(SDTProbesGNULinuxTest.java:76)
at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
```

The reason is that the new test now enables `DTraceMethodProbes`, `DTraceAllocProbes`, `DTraceMonitorProbes`, but VM never actually checks those options are available. So test thinks dtrace support is there, yet there are no relevant sections in `libjvm.so`, so the test fails. We should extend the argument checks to those three options too.

Unfortunately, there are tests that do `DTrace*` flags, we should make them sense the build capability, and act accordingly. This simplifies some existing tests too.

Additional testing:
 - [x] Linux x86_64 fastdebug (dtrace enabled), affected test still passes
 - [x] Linux x86_32 fastdebug (dtrace disabled due to missing dependencies), affected test now passes

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281822](https://bugs.openjdk.java.net/browse/JDK-8281822): Test failures on non-DTrace builds due to incomplete DTrace* flags handling

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7477/head:pull/7477` \
`$ git checkout pull/7477`

Update a local copy of the PR: \
`$ git checkout pull/7477` \
`$ git pull https://git.openjdk.java.net/jdk pull/7477/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7477`

View PR using the GUI difftool: \
`$ git pr show -t 7477`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7477.diff">https://git.openjdk.java.net/jdk/pull/7477.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Looks good!

So JDK-8278423 exposed this pre-existing bug.

Thanks,
David

**Comment by shipilev** [comment]
> So JDK-8278423 exposed this pre-existing bug.

Yup!

**Comment by shipilev** [comment]
> > So JDK-8278423 exposed this pre-existing bug.
> 
> Yup!

Come to think about it, I can rename this bug to something more generic and backport it, so that older VMs barf on trying to enable DTrace where not available.

**Comment by shipilev** [comment]
All right, there are legitimate GHA test failures now in `compiler/runtime/Test8168712.java`, which enables `-XX:+DTraceMethodProbes`. I'll fix that too...

**Comment by shipilev** [comment]
...and done. @dholmes-ora, you might want to take a look again, as went much farther than the initial PR.

File: test/hotspot/jtreg/serviceability/dtrace/DTraceOptionsTest.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/dtrace/DTraceOptionsTest.java:28]: macOS should be a possibility too. But if you check explicitly for hasDTrace then you don't also need to limit the platform.

**Review by dholmes-ora — APPROVED** [review]
Yeah that is certainly a lot more changes, the main value of which seems to be for backports given we're even thinking of dropping DTrace support entirely. (I decided not to bother trying to add a "hasDTrace" capability when fixing the issue with the VMDeprecatedOptions test as it was not worth the time and effort IMO).

One query about restricting the test to Linux below - and I'll need to take the final version for a spin in our CI before pushing, but tacit approval.

Thanks,
David

File: test/hotspot/jtreg/serviceability/dtrace/DTraceOptionsTest.java
Comment (shipilev) [reply test/hotspot/jtreg/serviceability/dtrace/DTraceOptionsTest.java]: Right, fixed.

**Comment by shipilev** [comment]
> One query about restricting the test to Linux below - and I'll need to take the final version for a spin in our CI before pushing, but tacit approval.

OK, I am seeing another GHA failure, which is due to another instance `Whitebox` missing a method declaration. I pushed the version that should fix that failure now. Once GHA is green, I'd say the fix is final.

**Comment by dholmes-ora** [comment]
I will run it through our CI and report back.

File: test/hotspot/jtreg/compiler/runtime/Test8168712.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/compiler/runtime/Test8168712.java:27]: DTrace is not critical to this test. IIUC the original fix and test had an issue when combined with DTrace hence the test enabled it to check it was fixed by a further patch. So we actually want to run this test regardless. In which case we need to add another test section for when `!vm.hasDTrace` that doesn't pass the DTrace flag.

File: src/hotspot/share/runtime/arguments.cpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/arguments.cpp:2911]: Looking at this again, in light of the test changes, I have to wonder if the pre-existing treatment of `ExtendedDTraceProbes` was wrong and we should simply have used `UNSUPPORTED_OPTION` in these cases?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Test change requested - see below.

Also another thought on the fix below.

File: src/hotspot/share/runtime/arguments.cpp
Comment (shipilev) [reply src/hotspot/share/runtime/arguments.cpp:2911]: Maybe, but this block also handles ergonomic opt-in to `DTrace*` flags when `ExtendedDTraceProbe` is enabled, and it matches the style of the similar checks. I'd leave it as is.

File: test/hotspot/jtreg/compiler/runtime/Test8168712.java
Comment (shipilev) [reply test/hotspot/jtreg/compiler/runtime/Test8168712.java:27]: Right. Added, see new commit.

File: src/hotspot/share/runtime/arguments.cpp
Comment (dholmes-ora) [reply src/hotspot/share/runtime/arguments.cpp:2911]: ok

**Review by dholmes-ora — APPROVED** [review]
Looks good.

Still waiting for test results - no failures but I need to confirm new tests actually ran. :)

**Comment by dholmes-ora** [comment]
Tests are fine. Thanks.

**Comment by shipilev** [comment]
OK, cool. GHA are also green. I'll wait for another reviewer...

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by shipilev** [comment]
Thanks!

/integrate

**Pull request closed without merging**
